### PR TITLE
Azure library migration - Namespace flows - Part 1/4

### PR DESCRIPTION
**Pull request by romayalon**
Signed-off-by: Romy <romy2232@gmail.com>

### Explain the changes
1. this PR should be merged only after namespace tests PR https://github.com/noobaa/noobaa-core/pull/6743 will be completed and merged.
2. Required the new Azure library (https://github.com/Azure/azure-sdk-for-js) and created new wrap for the library.
3. Changed Namespace blob, namespace monitor and object_sdk with the needed changes.

### Issues: Fixed #xxx / Gap #xxx
1. 

### Testing Instructions:
1.

### CODE REVIEW

File: src/sdk/namespace_blob.js [hunk 1]
```diff
@@ -630,18 +609,21 @@ class NamespaceBlob {
...
+        // list objects returns properties
+        // head object returns flat
+        const flat_obj = obj.properties || obj;
+        const blob_etag = blob_utils.parse_etag(flat_obj.etag);
+        const md5_b64 = flat_obj.contentMD5 || '';
         const md5_hex = Buffer.from(md5_b64, 'base64').toString('hex');
         const etag = md5_hex || blob_etag;
-        const size = parseInt(obj.contentLength, 10);
+        let size;
+        size = parseInt(flat_obj.contentLength, 10);
```
Comment (liranmauda) [inline src/sdk/namespace_blob.js:626]:
```suggestion
       const size = parseInt(flat_obj.contentLength, 10);
```

File: src/util/cloud_utils.js [hunk 2]
```diff
@@ -64,6 +64,22 @@ function get_azure_connection_string(params) {
     return connection_string;
 }
 
+function get_azure_new_connection_string(params) {
+    let endpoint_url = url.parse(params.endpoint);
```
Comment (liranmauda) [inline src/util/cloud_utils.js:68]:
```suggestion
    const endpoint_url = url.parse(params.endpoint);
```

File: src/util/cloud_utils.js [hunk 3]
```diff
@@ -64,6 +64,22 @@ function get_azure_connection_string(params) {
     return connection_string;
 }
 
+function get_azure_new_connection_string(params) {
```
Comment (liranmauda) [inline src/util/cloud_utils.js:67]: Let's add a TODO that when there will no need for both functions (`get_azure_connection_string` and 
`get_azure_new_connection_string` ) we can remove the old and rename the new

File: src/util/new_azure_storage_wrap.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
+/* Copyright (C) 2016 NooBaa */
+'use strict';
+var crypto = require("crypto");
```
Comment (liranmauda) [inline src/util/new_azure_storage_wrap.js:3]:
```suggestion
const crypto = require("crypto");
```

**Review by liranmauda — CHANGES_REQUESTED** [review]

File: src/sdk/namespace_blob.js [hunk 5]
```diff
@@ -8,31 +8,38 @@ const crypto = require('crypto');
...
 
-    constructor({ namespace_resource_id, rpc_client, connection_string, container, account_name }) {
+    constructor({ namespace_resource_id, rpc_client, connection_string, container, account_name, account_key }) {
         this.namespace_resource_id = namespace_resource_id;
         this.connection_string = connection_string;
         this.container = container;
-        this.blob = azure_storage.createBlobService(connection_string);
+        this.blob = azure_storage.BlobServiceClient.fromConnectionString(connection_string);
+        // needed only for generateBlockIdPrefix() and get_block_id() functions
+        this.old_blob = azure_storage.get_old_blob_service_conn_string(connection_string);
```
[the remark this answers is not in the document]
Comment (romayalon) [reply src/sdk/namespace_blob.js]: Added

File: src/sdk/namespace_blob.js [hunk 6]
```diff
@@ -56,33 +63,30 @@ class NamespaceBlob {
...
-            this.container,
-            inspect(params),
-            'blobs', inspect(blobs),
-            'dirs', inspect(dirs)
-        );
+        const max_keys = params.limit || 1000;
+        let iterator;
+        if (params.delimiter) {
+            iterator = await this.container_client.listBlobsByHierarchy(
+                params.delimiter, { prefix: params.prefix }).byPage({ continuationToken: params.key_marker, maxPageSize: max_keys});
```
[the remark this answers is not in the document]
Comment (romayalon) [reply src/sdk/namespace_blob.js:77]: yes

File: src/sdk/namespace_blob.js [hunk 7]
```diff
@@ -56,33 +63,30 @@ class NamespaceBlob {
...
+        console.log('list_objects: ', response.continuationToken, response.segment.blobItems, response.segment.blobPrefixes);
 
         return {
-            objects: _.map(blobs, obj => this._get_blob_object_info(obj, params.bucket)),
-            common_prefixes: _.map(dirs, dir => dir.Name[0]),
-            is_truncated: Boolean(next_marker),
-            next_marker
+            objects: _.map(response.segment.blobItems, obj => this._get_blob_object_info(obj, params.bucket)),
+            common_prefixes: _.map(response.segment.blobPrefixes, prefix => prefix.name),
+            is_truncated: Boolean(response.continuationToken),
```
[the remark this answers is not in the document]
Comment (romayalon) [reply src/sdk/namespace_blob.js:88]: just continutationToken, so it's fine

File: src/sdk/namespace_blob.js [hunk 8]
```diff
@@ -316,26 +305,24 @@ class NamespaceBlob {
...
-                callback)
+        const blob_client = this._get_blob_client(params.key);
+        await blob_client.stageBlock(
+            params.block_id,
+            () => params.source_stream.pipe(count_stream),
+            params.size // streamLength really required ???
         );
     }
 
     async commit_blob_block_list(params) {
```
[the remark this answers is not in the document]
Comment (romayalon) [reply src/sdk/namespace_blob.js:316]: @nimrod-becker @liranmauda we have these similar implementations in many other namespaces,
removing it from here is not enough,
shall I open a github issue so we will remove all the flow?

File: src/sdk/namespace_blob.js [hunk 9]
```diff
@@ -316,26 +305,24 @@ class NamespaceBlob {
...
-        return P.fromCallback(callback => this.blob.commitBlocks(params.bucket, params.key, block_list, callback));
+        //const is_mix = _.uniqBy(params.block_list, 'type').length > 1;
+        //const list_to_use = is_mix ? 'LatestBlocks' : MAP_BLOCK_LIST_TYPE[params.block_list[0].type];
+        // const block_list = {};
+        //block_list[list_to_use] = params.block_list.map(block => block.block_id);
+        const blob_client = this._get_blob_client(params.key);
+        return blob_client.commitBlockList(params.block_list);
     }
 
     async get_blob_block_lists(params) {
```
[the remark this answers is not in the document]
Comment (romayalon) [reply src/sdk/namespace_blob.js:328]: I commented on the other comment, resolving

**Review by liranmauda — APPROVED** [review]
LGTM

**Pull request merged**